pyccel/parser/semantic.py

```diff
@@ -311,27 +276,16 @@ def check_for_variable(self, name):
                 if i.name == name:
                     var = i
                     return var
-
-        # Walk up nested loops (if any)
-        container = self.namespace
-        while container.is_loop:
-            container = container.parent_scope
-
-        # Walk up the tree of Scope objects, until the root if needed
-        while container:
-            var = self._get_variable_from_scope(name, container)
-            if var is not None:
-                return var
-            container = container.parent_scope
-
-        return None
+            return None
+        else:
+            return self.namespace.find_in_scope(name, 'variables')
 
     def get_variable(self, name):
         """ Like 'check_for_variable', but raise Pyccel error if Variable is not found.
         """
         var = self.check_for_variable(name)
         if var is None:
-            errors.report(UNDEFINED_VARIABLE, symbol=name,
+            return errors.report(UNDEFINED_VARIABLE, symbol=name,
```

Comment:
Pylint raises an error:
> Either all return statements in a function should return an expression, or none of them should.

This can be fixed by adding either a `return` or a `raise` before `errors.report`. `return` seems safer to me because `raise` interrupts the flow, which `errors.report` already does internally

Reply:
This seems to be related to [this issue](https://github.com/PyCQA/pylint/issues/1782) and [this PR](https://github.com/PyCQA/pylint/pull/4304). So we should add a `NoReturn` type hint somewhere...

Reply:
It should be sufficient to declare `def report(self, ...) -> typing.NoReturn`. Please see [this test file](https://github.com/hippo91/pylint/blob/main/tests/functional/i/inconsistent/inconsistent_returns_noreturn.py) in the pylint repository.

Reply:
Actually I am mistaken, here (and in most other situations) we use `errors.report` to raise an exception, but it actually returns `None` in these two cases:

1.  `blocker = False`
2. `(self.mode == 'user') and (severity == 'internal')`

Hence we cannot say that this function never returns. Maybe our error report functions should be improved in the future? I think that your current solution is the most sensible for now.

Reply:
Yes indeed. This only raises an exception if `blocker == True` or `severity = 'fatal'`